### PR TITLE
Support methods on generic types.

### PR DESCRIPTION
**Pull request by nevkontakte**
Generic type methods are compiled largely in the same way as standalone generic functions, into a generic function factory. However, unlike a standalone function, the method factory is registered in the receiver type factory and is invoked when the type is being instantiated, with type arguments passed to the generic type. The instantiated method is then attached to the type instance.

At this point we have we are down to 52 failing test cases, and looking through them reveals a few bugs in this PR's implementation, however I will address those separately.

Updates #1013.

/cc @paralin

### CODE REVIEW

**Comment by paralin** [comment]
lgtm, great work as always!

File: compiler/astutil/astutil.go [hunk 1]
```diff
@@ -69,6 +76,12 @@ func FuncKey(d *ast.FuncDecl) string {
 	if star, ok := recv.(*ast.StarExpr); ok {
 		recv = star.X
 	}
+	if index, ok := recv.(*ast.IndexExpr); ok {
+		recv = index.X
+	}
+	if index, ok := recv.(*ast.IndexListExpr); ok {
+		recv = index.X
+	}
```
Comment (flimzy) [inline compiler/astutil/astutil.go:84]: Could a type switch be used here for brevity and clarity?

File: compiler/package.go [hunk 2]
```diff
@@ -530,7 +566,7 @@ func translateFunction(typ *ast.FuncType, recv *ast.Ident, body *ast.BlockStmt,
 		// If a blocking function is being resumed, initialize local variables from the saved context.
 		localVarDefs = fmt.Sprintf("var {%s, $c} = $restore(this, {%s});\n", strings.Join(localVars, ", "), strings.Join(params, ", "))
 		// If the function gets blocked, save local variables for future.
-		saveContext := fmt.Sprintf("var $f = {$blk: "+funcRef+", $c: true, $r, %s};", strings.Join(c.localVars, ", "))
+		saveContext := fmt.Sprintf("var $f = {$blk: "+instanceVar+", $c: true, $r, %s};", strings.Join(c.localVars, ", "))
```
Comment (flimzy) [inline compiler/package.go:569]: nit: although this predates this PR, building a format string with string concatenation seems a bit odd, given that `instanceVar` doesn't contain formatting verbs.  :shrug:

**Review by flimzy — APPROVED** [review]
a couple optional comments inline. Otherwise looks good to me!

File: compiler/astutil/astutil.go (hunk 1 above)
Comment (nevkontakte) [reply compiler/astutil/astutil.go:84]: Done.

File: compiler/package.go (hunk 2 above)
Comment (nevkontakte) [reply compiler/package.go:569]: Good catch, done.

**Pull request merged**